File: pkg-py/tests/test_remote_chembl.py
"""ChEMBL remote mode: live existence checks via the ChEMBL id-lookup endpoint."""

import pytest

import biobouncer
import biobouncer._remote as remote
from biobouncer import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def test_url_builds_the_chembl_id_lookup_endpoint():
    from biobouncer._registry import get_source

    url = remote._chembl_url(get_source("chembl"), "CHEMBL25")
    assert url == (
        "https://www.ebi.ac.uk/chembl/api/data/chembl_id_lookup/CHEMBL25.json"
    )


def test_existing_entity_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biobouncer.check_id("CHEMBL25", source_db="chembl", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "CHEMBL25"
    assert res.suggestion is None


def test_absent_entity_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(404))
    res = biobouncer.check_id("CHEMBL99999999", source_db="chembl", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_lowercase_suggests_the_uppercase_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biobouncer.check_id("chembl25", source_db="chembl", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "CHEMBL25"


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biobouncer.check_id("CHEMBLABC", source_db="chembl", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biobouncer.check_id("CHEMBL25", source_db="chembl", how="remote")
    assert not remote._remote_cache_path("chembl", "lookup", "CHEMBL25").is_file()

File: pkg-r/tests/testthat/test-remote-chembl.R
# Offline tests for chembl remote mode. Existence is checked against the ChEMBL
# id-lookup endpoint; the biobouncer.remote_transport option replaces it.

.stub_chembl <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}

test_that("the url addresses the chembl id-lookup endpoint", {
  resolver <- .remote_resolvers$chembl
  url <- resolver$url(NULL, "CHEMBL25")
  expect_identical(
    url,
    "https://www.ebi.ac.uk/chembl/api/data/chembl_id_lookup/CHEMBL25.json"
  )
})

test_that("an existing entity is valid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_chembl(200))
  res <- check_id("CHEMBL25", "chembl", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "CHEMBL25")
  expect_true(is.na(res$suggestion))
})

test_that("an absent entity is invalid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_chembl(404))
  res <- check_id("CHEMBL99999999", "chembl", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a lowercase id suggests the uppercase form", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_chembl(200))
  res <- check_id("chembl25", "chembl", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "CHEMBL25")
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("CHEMBLABC", "chembl", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_chembl(500))
  expect_error(
    check_id("CHEMBL25", "chembl", how = "remote"),
    class = "biobouncer_error_remote"
  )
  expect_false(file.exists(.remote_cache_path("chembl", "lookup", "CHEMBL25")))
})
